apps/tc/elpi/compiler.elpi

```diff
@@ -0,0 +1,234 @@
+  N > 0,
+  N1 is N - 1,
+  pi name ty f gen\ remove-ho-unification IsHead ff N1 Ty AppInst Premises LT 
+    [name | NameL] [ty | TypeL] [f | FunL] 
+    [gen | GenL] Fixes (Clause ty f name gen).
+
+pred pattern-fragment? i:term.
+pattern-fragment? (app [HD|TL]) :- 
+  not (HD = global _), distinct_names [HD | TL].
+
+pred get-pattern-fragment i:term, o:list term.
+get-pattern-fragment T1 TL :- !,
+  (pi L G GL\ fold-map (app L as G) GL G [G | GL] :- distinct_names L) =>
+  % (pi G GL\ fold-map (app _ as G) GL G GL) =>
+  (pi G GL\ fold-map (prod _ _ _ as G) GL G GL) => 
+  fold-map T1 [] _ TL.
+
+/*
+compile-aux [Ty Inst Premises PiAccRev UnivL IsPositive Clause No-Premises]
+Ty          : the type of the instance
+Inst        : the instance term on the form (global InstGref)
+Premises    : list of constraints/premises of an instances found from its type
+PiAccRev    : the list of pi variables accumulated from the (prod _ _ Bo) of the 
+              type of Inst. The will be used on the solution of the clause
+              coq.mk-app Inst {std.rev PiAccRev} Sol
+UnivL       : the list of universes of types inside Ty that are replaced with 
+              a fresh variable to simplify unification
+IsPositive  : bring the information about the positivity of the current sub-term
+              e.g. if T = A -> (B -> C) -> D, then:
+              D is positive in T, (B -> C) is negative, A is positive in T
+              C is positivie in (B -> C), B is negative in (B -> C)
+              IsPositive is used to know how to accumulate sigma Xi\ and pi x\ in the 
+              current clause 
+IsHead      : a prop [true|false] to know if we are compiling the head of the 
+              instance or one hypothesis of its premises
+Clause      : the solution to be returned
+No-Premises : a boolean saying if the returned clause as no premises that is an 
+              instance with no hypothesis
+*/
+pred compile-aux i:term, i:term, i:list prop, i:list term, i:list univ, i:bool, i:prop, o:prop, o:bool.
+:name "compiler-aux:start"
+compile-aux Ty I [] [] [X | XS] IsPositive IsHead (pi x\ C x) IsLeaf :- !,
+  pi x\ copy (sort (typ X)) (sort (typ x)) => copy Ty (Ty1 x),
+    compile-aux (Ty1 x) I [] [] XS IsPositive IsHead (C x) IsLeaf.
+compile-aux (prod N T F) I RevPremises ListVar [] IsPositive IsHead Clause ff :- !,
+  (if (IsPositive = tt) (Clause = pi x\ C x) (Clause = (pi x\ decl x N T => C x))),
+  pi p\ sigma NewPremise TC L\ 
+    if (get-TC-of-inst-type T TC, coq.TC.class? TC /*, not (occurs p (F p))*/) 
+      (compile-aux T p [] [] [] {neg IsPositive} false NewPremise _,
+      if (classes TC deterministic) 
+        (L = [do-once NewPremise | RevPremises])
+        (L = [NewPremise | RevPremises])) (L = RevPremises),
+    compile-aux (F p) I L [p | ListVar] [] IsPositive IsHead (C p) _.
+:if "simple-compiler"
+compile-aux Ty I RevPremises ListVar [] _ IsHead Clause tt :- !,
```

Comment:
change into a coq option

Reply:
off by default,
if a pattern fragment is found and the option is off, spit a warning